src/hotspot/share/memory/metaspaceShared.cpp

```diff
@@ -776,6 +777,132 @@ bool MetaspaceShared::try_link_class(InstanceKlass* ik, TRAPS) {
   }
 }
 
+void append_strings(char* buffer, size_t buff_len,  const char* arg) {
+  char* start = buffer + strlen(buffer);
+  snprintf(start, buff_len, "%s ", arg);
+}
+
+void MetaspaceShared::cmd_dump_shared_archive(outputStream* output, const char* cmd, const char* file_name,  TRAPS) {
+  // The existing file will be overwritten.
+  char filename[JVM_MAXPATHLEN];
+  const char* file = file_name;
+  assert(strcmp(cmd, "static_dump") == 0 || strcmp(cmd, "dynamic_dump") == 0, "Sanity check");
+  bool is_static = strcmp(cmd, "static_dump") == 0;
+  if (is_static) {
+    output->print_cr("Static dump");
+    if (file_name ==nullptr) {
+      os::snprintf(filename, sizeof(filename), "java_pid%d_static.jsa", os::current_process_id());
+      file = filename;
+    } else {
+      if (strstr(file_name, ".jsa") == nullptr) {
+        os::snprintf(filename, sizeof(filename), "%s.jsa", file_name);
+        file = filename;
+      }
+    }
```

Comment:
This block of code is very similar to lines 813 - 821 below. Maybe factor it into another function?

Reply:
changed to use java to dump. That will be much simple to deal with string.